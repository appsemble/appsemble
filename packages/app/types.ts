import { type EventEmitter } from 'events';

import { type ShowMessage } from '@appsemble/react-components';
import {
  type ActionDefinition,
  type AppDefinition,
  type BlockDefinition,
  type BlockManifest,
  type ProjectImplementations,
  type Remapper,
  type TeamMember,
  type UserInfo,
} from '@appsemble/types';
import { type AppConfigEntryGetter, type MessageGetter } from '@appsemble/utils';
import { type IconName } from '@fortawesome/fontawesome-common-types';
import { type Dispatch } from 'react';
import { type NavigateFunction, type Params } from 'react-router-dom';
import { type JsonValue } from 'type-fest';

import { type ActionCreators } from './utils/actions/index.js';
import { type AppStorage } from './utils/storage.js';

declare module '@appsemble/sdk' {
  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface Actions {
    [K: string]: ActionDefinition;
  }

  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface EventEmitters {
    [K: string]: never;
  }

  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface EventListeners {
    [K: string]: never;
  }

  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface Parameters {
    [K: string]: JsonValue;
  }

  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface Messages {
    [K: string]: Record<string, string>;
  }
}

export interface User extends UserInfo {
  scope: string;
}

declare global {
  interface Window {
    /**
     * This boolean indicates if Appsemble has loaded normally.
     *
     * If this is not been set, this means Appsemble didn’t load, probably because it doesn’t
     * recognize newer JavaScript features.
     */
    appsembleHasLoaded: boolean;

    settings: {
      apiUrl: string;
      appControllerCode: string;
      appControllerImplementations: ProjectImplementations;
      blockManifests: BlockManifest[];
      pageManifests: ProjectImplementations;
      vapidPublicKey: string;
      id: number;
      definition: AppDefinition;
      demoMode: boolean;
      languages: string[];
      logins: {
        icon: IconName;
        id: number;
        name: string;
        type: 'oauth2' | 'saml';
      }[];
      sentryDsn: string;
      sentryEnvironment: string;
      showAppsembleLogin: boolean;
      showAppsembleOAuth2Login: boolean;
      enableSelfRegistration: boolean;
      showDemoLogin: boolean;
      appUpdated: string;
      development: boolean;
    };
  }
}

export interface ShowDialogParams {
  actionCreators: ActionCreators;
  blocks: BlockDefinition[];
  closable?: boolean;
  data: any;
  close: () => void;
  fullscreen: boolean;
  prefix: string;
  prefixIndex: string;
  title?: Remapper;
}

export type ShowDialogAction = (params: ShowDialogParams) => () => void;
export interface ShowShareDialogParams {
  url?: string;
  text?: string;
  title?: string;
}
export type ShowShareDialog = (params: ShowShareDialogParams) => Promise<void>;

export interface FlowActions {
  back: (data: any) => Promise<any>;
  cancel: (data: any) => Promise<any>;
  finish: (data: any) => Promise<any>;
  next: (data: any) => Promise<any>;
  to: (data: any, step: string) => Promise<any>;
}

export type UpdateTeam = (team: TeamMember) => void;

export interface MakeActionParameters<D extends ActionDefinition> {
  app: AppDefinition;
  getAppMessage?: MessageGetter;
  getAppVariable?: AppConfigEntryGetter;
  appStorage: AppStorage;
  definition: D;
  extraCreators?: ActionCreators;
  flowActions?: FlowActions;
  navigate?: NavigateFunction;
  pageReady: Promise<void>;
  params: Readonly<Params<string>>;
  showShareDialog?: ShowShareDialog;
  showDialog?: ShowDialogAction;
  prefix: string;
  prefixIndex: string;
  pushNotifications: ServiceWorkerRegistrationContextType;
  ee: EventEmitter;
  remap: (remapper: Remapper, data: any, context?: Record<string, any>) => any;
  showMessage?: ShowMessage;
  getUserInfo: () => UserInfo;
  passwordLogin: (params: { username: string; password: string }) => Promise<void>;
  passwordLogout: () => Promise<void>;
  setUserInfo: Dispatch<UserInfo>;
  updateTeam: UpdateTeam;
  teams: TeamMember[];
  refetchDemoAppMembers: () => Promise<void>;
}

export type Permission = NotificationPermission | 'pending';

export interface ServiceWorkerRegistrationContextType {
  subscribe: () => Promise<PushSubscription>;
  unsubscribe: () => Promise<boolean>;
  permission: Permission;
  subscription: PushSubscription;
  requestPermission: () => Promise<NotificationPermission>;
}
