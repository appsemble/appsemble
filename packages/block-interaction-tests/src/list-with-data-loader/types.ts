import '@appsemble/data-loader/block.js';
import '@appsemble/list/block.js';
