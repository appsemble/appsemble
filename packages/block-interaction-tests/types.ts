declare module '@appsemble/sdk' {
  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface EventEmitters {
    [K: string]: never;
  }

  // eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
  interface EventListeners {
    [K: string]: never;
  }
}
