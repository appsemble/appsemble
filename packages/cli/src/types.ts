import { AppVisibility } from '@appsemble/types';
import { PackageJson } from 'read-pkg-up';
import { JsonObject } from 'type-fest';

/**
 * The base arguments from the command line.
 *
 * See {@link ./index.js}.
 */
export interface BaseArguments {
  verbose?: number;
  quiet?: number;
  remote: string;
  clientCredentials?: string;
}

/**
 * The arguments passed to `appsemble app update`.
 *
 * See {@link ./commands/app/update.js}.
 */
export interface UpdateAppArguments extends BaseArguments {
  /**
   * The ID of the app to update.
   */
  appId: number;

  /**
   * The path in which the app YAML is located.
   */
  path: string;

  /**
   * Visibility of the app in the public app store.
   */
  visibility: AppVisibility;

  /**
   * Whether the app should be marked as a template.
   */
  template: boolean;

  /**
   * Whether the locked property should be ignored.
   */
  force: boolean;
}

export interface MonoRepoPackageJson extends PackageJson {
  appsembleServer?: JsonObject;
}

export interface AppsembleContext {
  /**
   * If `remote` is specified, this will override `--remote` passed by the command line.
   */
  remote?: string;

  /**
   * If `organization` is specified, this will override `--organization` passed by the command line
   * when creating apps.
   */
  organization?: string;

  icon?: string;

  iconBackground?: string;

  maskableIcon?: string;

  /**
   * If `id` is specified, this will override `--id` passed by the command line when
   * updating an app.
   */
  id?: number;

  /**
   * The Google Analytics ID that should be used for the app.
   */
  googleAnalyticsId?: string;

  /**
   * The custom Sentry DSN for the app.
   */
  sentryDsn?: string;

  /**
   * The environment for the custom Sentry DSN for the app.
   */
  sentryEnvironment?: string;

  /**
   * If `visibility` is specified, this will override `--visibility` passed on the command line.
   */
  visibility?: AppVisibility;

  /**
   * If `template` is specified, this will override `--template` passed on the command line.
   */
  template?: boolean;
}

export interface AppsembleRC {
  /**
   * The background color to use for maskable icons.
   */
  iconBackground?: string;

  context?: Record<string, AppsembleContext>;
}
