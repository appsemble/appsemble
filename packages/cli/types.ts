import { type AppVisibility } from '@appsemble/types';
import { type PackageJson } from 'read-package-up';
import { type JsonObject } from 'type-fest';

/**
 * The base arguments from the command line.
 *
 * See {@link ./index.js}.
 */
export interface BaseArguments {
  verbose?: number;
  quiet?: number;
  remote: string;
  clientCredentials?: string;
}

/**
 * The arguments passed to `appsemble app update`.
 *
 * See {@link ./commands/app/update.js}.
 */
export interface UpdateAppArguments extends BaseArguments {
  /**
   * The ID of the app to update.
   */
  appId: number;

  /**
   * The path in which the app YAML is located.
   */
  path: string;

  /**
   * Visibility of the app in the public app store.
   */
  visibility: AppVisibility;

  /**
   * Whether the app should be marked as a template.
   */
  template: boolean;

  /**
   * Whether the locked property should be ignored.
   */
  force: boolean;
}

// @ts-expect-error Messed up
export interface MonoRepoPackageJson extends PackageJson {
  // @ts-expect-error Messed up
  appsembleServer?: JsonObject;
}
