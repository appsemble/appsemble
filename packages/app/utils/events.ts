import { type EventEmitter } from 'events';

import { type BlockDefinition } from '@appsemble/lang-sdk';
import { type Events } from '@appsemble/sdk';
import { type ProjectManifest } from '@appsemble/types';
import { has } from '@appsemble/utils';
import { addBreadcrumb } from '@sentry/browser';

/**
 * Create the events object that is passed to a block.
 *
 * @param ee The internal event emitter to use.
 * @param ready A promise to wait for before emitting any events.
 * @param manifest The block manifest.
 * @param definition The block definition.
 * @returns An events object that may be passed into a block.
 */
export function createEvents(
  ee: EventEmitter,
  ready: Promise<void>,
  manifest?: ProjectManifest['events'],
  definition?: BlockDefinition['events'],
): Events {
  function createProxy<
    E extends keyof Events,
    M extends keyof NonNullable<ProjectManifest['events']>,
  >(manifestKey: M, createFn: (registered: boolean, key: string) => Events[E][string]): Events[E] {
    return new Proxy<Events[E]>(
      {},
      {
        get(target, key) {
          if (typeof key !== 'string') {
            return;
          }
          if (has(target, key)) {
            return target[key];
          }
          if (!has(manifest?.[manifestKey], key) && !has(manifest?.[manifestKey], '$any')) {
            return;
          }
          const handler: Events[E][string] = createFn(has(definition?.[manifestKey], key), key);
          // eslint-disable-next-line no-param-reassign
          target[key as keyof Events[E]] = handler;
          return handler;
        },
      },
    );
  }

  const emit = createProxy<'emit', 'emit'>('emit', (implemented, key) =>
    implemented
      ? async (data, error) => {
          await ready;
          const name = definition?.emit?.[key];
          // @ts-expect-error 2345 argument of type is not assignable to parameter of type
          // (strictNullChecks)
          ee.emit(name, data, error === '' ? 'Error' : error);
          addBreadcrumb({
            category: 'appsemble.event',
            // @ts-expect-error 2345 argument of type is not assignable to parameter of type
            // (strictNullChecks)
            data: { name, listeners: String(ee.listenerCount(name)) },
          });
          return true;
        }
      : // eslint-disable-next-line require-await
        async () => false,
  );

  const on = createProxy<'on', 'listen'>('listen', (implemented, key) =>
    implemented
      ? (callback) => {
          // @ts-expect-error 2345 argument of type is not assignable to parameter of type
          // (strictNullChecks)
          ee.on(definition.listen?.[key], callback);
          return true;
        }
      : () => false,
  );

  const off = createProxy<'off', 'listen'>('listen', (implemented, key) =>
    implemented
      ? (callback) => {
          // @ts-expect-error 2345 argument of type is not assignable to parameter of type
          // (strictNullChecks)
          ee.off(definition.listen?.[key], callback);
          return true;
        }
      : () => false,
  );

  return { emit, on, off };
}
