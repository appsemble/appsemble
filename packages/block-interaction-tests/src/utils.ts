import { type EventEmitter } from 'node:events';

import {
  ActionError,
  type BlockDefinition,
  type ProjectManifest,
  remap,
  type Remapper,
} from '@appsemble/lang-sdk';
import {
  type BlockUtils,
  type BootstrapParams,
  type Events,
  type Messages,
  type Theme,
} from '@appsemble/sdk';
import { defaultLocale, has } from '@appsemble/utils';

export function remapWithContext(
  remapper: Remapper,
  data: any,
  context?: Record<string, any>,
): unknown {
  return remap(remapper, data, {
    // @ts-expect-error strictNullChecks not assignable to type
    getMessage: null,
    // @ts-expect-error strictNullChecks not assignable to type
    getVariable: null,
    appId: 1,
    url: 'https://example.com/en/example',
    appUrl: 'https://example.com',
    // @ts-expect-error strictNullChecks not assignable to type
    appMemberInfo: null,
    // @ts-expect-error strictNullChecks not assignable to type
    context,
    locale: defaultLocale,
  });
}

export function getDefaultUtils(): BlockUtils {
  return {
    showMessage(message) {
      return message;
    },
    formatMessage<T extends keyof Messages>(
      message: T,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      ...args: Messages[T] extends never ? [] : [Messages[T]]
    ) {
      return message;
    },
    asset(assetId) {
      return assetId;
    },
    fa(icon) {
      return icon;
    },
    // @ts-expect-error strictNullChecks not assignable to type
    remap: remapWithContext,
    isMobile: window.innerWidth < 768,
    menu() {
      return null;
    },
    isActionError(input): input is ActionError {
      return input instanceof ActionError;
    },
    addCleanup() {
      return null;
    },
  };
}

export function getDefaultBootstrapParams(): Pick<
  BootstrapParams,
  'data' | 'events' | 'path' | 'pathIndex' | 'shadowRoot' | 'theme' | 'utils'
> & { ready: () => Promise<void> } {
  return {
    data: {},
    path: '',
    pathIndex: '',
    shadowRoot: document?.createElement('div').attachShadow({ mode: 'open' }),
    theme: {} as Theme,
    utils: getDefaultUtils(),
    events: {
      // @ts-expect-error strictNullChecks not assignable to type
      emit: null,
      // @ts-expect-error strictNullChecks not assignable to type
      on: null,
      // @ts-expect-error strictNullChecks not assignable to type
      off: null,
    },
    ready() {
      return Promise.resolve();
    },
  };
}

/**
 * Create the events object that is passed to a block.
 *
 * FIXME This function is the same as @appsemble/app/utils/createEvents
 *
 * @param ee The internal event emitter to use.
 * @param ready A promise to wait for before emitting any events.
 * @param manifest The block manifest.
 * @param definition The block definition.
 * @returns An events object that may be passed into a block.
 */
export function createEvents(
  ee: EventEmitter,
  ready: Promise<void>,
  manifest?: ProjectManifest['events'],
  definition?: BlockDefinition['events'],
): Events {
  function createProxy<
    E extends keyof Events,
    M extends keyof NonNullable<ProjectManifest['events']>,
  >(
    manifestKey: M,
    createFn: (registered: boolean, key: string) => Events[E][keyof Events[E]],
  ): Events[E] {
    return new Proxy<Events[E]>({} as Events[E], {
      get(target, key) {
        if (typeof key !== 'string') {
          return;
        }
        if (has(target, key)) {
          return target[key as keyof Events[E]];
        }
        if (!has(manifest?.[manifestKey], key) && !has(manifest?.[manifestKey], '$any')) {
          return;
        }
        const handler: Events[E][keyof Events[E]] = createFn(
          has(definition?.[manifestKey], key),
          key,
        );
        // eslint-disable-next-line no-param-reassign
        target[key as keyof Events[E]] = handler;
        return handler;
      },
    });
  }

  const emit = createProxy<'emit', 'emit'>('emit', (implemented, key) =>
    implemented
      ? async (data, error) => {
          await ready;
          const name = definition?.emit?.[key];
          // @ts-expect-error strictNullChecks not assignable to type
          ee.emit(name, data, error === '' ? 'Error' : error);
          return true;
        }
      : // eslint-disable-next-line require-await
        async () => false,
  );

  const on = createProxy<'on', 'listen'>('listen', (implemented, key) =>
    implemented
      ? (callback) => {
          // @ts-expect-error strictNullChecks not assignable to type
          ee.on(definition?.listen?.[key], callback);
          return true;
        }
      : () => false,
  );

  const off = createProxy<'off', 'listen'>('listen', (implemented, key) =>
    implemented
      ? (callback) => {
          // @ts-expect-error strictNullChecks not assignable to type
          ee.off(definition?.listen?.[key], callback);
          return true;
        }
      : () => false,
  );

  return { emit, on, off };
}
