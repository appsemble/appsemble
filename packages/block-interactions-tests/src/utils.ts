import { type EventEmitter } from 'node:events';

import {
  ActionError,
  type BlockDefinition,
  type ProjectManifest,
  remap,
  type Remapper,
} from '@appsemble/lang-sdk';
import {
  type BlockUtils,
  type BootstrapParams,
  type Events,
  type Messages,
  type Theme,
} from '@appsemble/sdk';
import { defaultLocale, has } from '@appsemble/utils';

export function remapWithContext(
  remapper: Remapper,
  data: any,
  context?: Record<string, any>,
): unknown {
  return remap(remapper, data, {
    getMessage: null,
    getVariable: null,
    appId: 1,
    url: 'https://example.com/en/example',
    appUrl: 'https://example.com',
    appMemberInfo: null,
    context,
    locale: defaultLocale,
  });
}

export function getDefaultUtils(): BlockUtils {
  return {
    showMessage(message) {
      return message;
    },
    formatMessage<T extends keyof Messages>(
      message: T,
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      ...args: Messages[T] extends never ? [] : [Messages[T]]
    ) {
      return message;
    },
    asset(assetId) {
      return assetId;
    },
    fa(icon) {
      return icon;
    },
    remap: remapWithContext,
    isMobile: window.innerWidth < 768,
    menu() {
      return null;
    },
    isActionError(input): input is ActionError {
      return input instanceof ActionError;
    },
    addCleanup() {
      return null;
    },
  };
}

export function getDefaultBootstrapParams(): Pick<
  BootstrapParams,
  'data' | 'path' | 'pathIndex' | 'shadowRoot' | 'theme' | 'utils'
> {
  return {
    data: {},
    path: '',
    pathIndex: '',
    shadowRoot: document?.createElement('div').attachShadow({ mode: 'open' }),
    theme: {} as Theme,
    utils: getDefaultUtils(),
  };
}

/**
 * Create the events object that is passed to a block.
 *
 * FIXME This function is the same as @appsemble/app/utils/createEvents
 *
 * @param ee The internal event emitter to use.
 * @param ready A promise to wait for before emitting any events.
 * @param manifest The block manifest.
 * @param definition The block definition.
 * @returns An events object that may be passed into a block.
 */
export function createEvents(
  ee: EventEmitter,
  ready: Promise<void>,
  manifest?: ProjectManifest['events'],
  definition?: BlockDefinition['events'],
): Events {
  function createProxy<
    E extends keyof Events,
    M extends keyof NonNullable<ProjectManifest['events']>,
  >(manifestKey: M, createFn: (registered: boolean, key: string) => Events[E][string]): Events[E] {
    return new Proxy<Events[E]>({} as any, {
      get(target, key) {
        if (typeof key !== 'string') {
          return;
        }
        if (has(target, key)) {
          return target[key];
        }
        if (!has(manifest?.[manifestKey], key) && !has(manifest?.[manifestKey], '$any')) {
          return;
        }
        const handler: Events[E][string] = createFn(has(definition?.[manifestKey], key), key);
        // eslint-disable-next-line no-param-reassign
        target[key as keyof Events[E]] = handler;
        return handler;
      },
    });
  }

  const emit = createProxy<'emit', 'emit'>('emit', (implemented, key) =>
    implemented
      ? async (data, error) => {
          await ready;
          const name = definition?.emit?.[key];
          ee.emit(name, data, error === '' ? 'Error' : error);
          return true;
        }
      : // eslint-disable-next-line require-await
        async () => false,
  );

  const on = createProxy<'on', 'listen'>('listen', (implemented, key) =>
    implemented
      ? (callback) => {
          ee.on(definition.listen?.[key], callback);
          return true;
        }
      : () => false,
  );

  const off = createProxy<'off', 'listen'>('listen', (implemented, key) =>
    implemented
      ? (callback) => {
          ee.off(definition.listen?.[key], callback);
          return true;
        }
      : () => false,
  );

  return { emit, on, off };
}
