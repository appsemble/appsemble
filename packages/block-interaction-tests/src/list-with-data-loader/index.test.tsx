import { EventEmitter } from 'node:events';

import { DataLoader } from '@appsemble/data-loader/src/bootstrap.js';
import { List } from '@appsemble/list/src/bootstrap.js';
import { type BlockProps, Context } from '@appsemble/preact';
import { render, screen } from '@testing-library/preact';
import { expect, it } from 'vitest';

import '../../types.js';
import { createEvents, getDefaultBootstrapParams } from '../utils.js';

const defaultBootstrapParams = getDefaultBootstrapParams();

it('should render a list with data from a data-loader', async () => {
  // eslint-disable-next-line unicorn/prefer-event-target
  const ee = new EventEmitter();

  let resolveReady: () => void;
  const ready = new Promise<void>((resolve) => {
    resolveReady = resolve;
  });

  const listProps: BlockProps = {
    ...defaultBootstrapParams,
    ready: resolveReady,
    parameters: {
      itemDefinition: {
        header: {
          title: 'test',
        },
      },
    },
    actions: {
      onClick: { type: 'noop' },
      onDrop: { type: 'noop' },
    } as any,
    events: createEvents(ee, ready, { listen: { data: {} } }, { listen: { data: 'data' } }),
  };

  DataLoader({
    ...defaultBootstrapParams,
    parameters: {} as any,
    actions: {
      onLoad: Object.assign(
        // eslint-disable-next-line func-names,prefer-arrow-callback
        function <R>(): Promise<R> {
          return Promise.resolve([{ id: 1 }, { id: 2 }]) as Promise<R>;
        },
        { type: 'noop' as const },
      ),
    } as any,
    events: createEvents(ee, ready, { emit: { data: {} } }, { emit: { data: 'data' } }),
  });

  render(
    <Context.Provider value={listProps}>
      <List dataTestId="test-list-with-data-loader" {...listProps} />
    </Context.Provider>,
  );

  // Wait for the data to load
  await new Promise((r) => {
    setTimeout(r, 100);
  });

  const testSelect = screen.getByTestId('test-list-with-data-loader');
  expect(testSelect).toMatchSnapshot();
});
