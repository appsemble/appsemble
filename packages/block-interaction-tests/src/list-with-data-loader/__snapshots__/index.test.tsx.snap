// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render a list with data from a data-loader 1`] = `
<div
  class="px-4"
  data-testid="test-list-with-data-loader"
>
  <ul
    class="py-4"
  >
    <li>
      <div
        class="is-flex is-align-items-center"
      >
        <div
          class=" listItem"
        >
          <div
            class="item has-text-left is-flex is-flex-direction-column my-1 py-2 pl-3 pr-2"
            id=".item.1"
          >
            <div
              class="headerWrapper is-flex"
              role="button"
              tabindex="-1"
            >
              <div
                class="is-flex image"
              >
                <div />
                <button
                  class="item has-text-left is-block"
                  type="button"
                >
                  <div
                    class="is-flex is-justify-content-space-between is-align-items-center"
                  >
                    <div>
                      <h4
                        class="title"
                      >
                        test
                      </h4>
                    </div>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
    <li>
      <div
        class="is-flex is-align-items-center"
      >
        <div
          class=" listItem"
        >
          <div
            class="item has-text-left is-flex is-flex-direction-column my-1 py-2 pl-3 pr-2"
            id=".item.2"
          >
            <div
              class="headerWrapper is-flex"
              role="button"
              tabindex="-1"
            >
              <div
                class="is-flex image"
              >
                <div />
                <button
                  class="item has-text-left is-block"
                  type="button"
                >
                  <div
                    class="is-flex is-justify-content-space-between is-align-items-center"
                  >
                    <div>
                      <h4
                        class="title"
                      >
                        test
                      </h4>
                    </div>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </li>
  </ul>
</div>
`;
